### PR TITLE
Housekeeping of env vars, defaults and annotations

### PR DESCRIPTION
**Pull request by xoen**
Possibly breaking 3rd party:
- [changed](https://github.com/ministryofjustice/analytics-platform-idler/commit/ed0d16ed5e0436eaa6cef5e9405a29c339db3a3d#diff-225132354aa2a9f4d7ecebb48833c06bR99) default for `LABEL_SELECTOR` from `app=rstudio` to `mojanalytics.xyz/idleable=true` which is what we use in the config (breaking if they don't set this env var and they rely on default)

Breaking:
- [renamed](https://github.com/ministryofjustice/analytics-platform-idler/commit/ed0d16ed5e0436eaa6cef5e9405a29c339db3a3d#diff-225132354aa2a9f4d7ecebb48833c06bR36) env var `RSTUDIO_ACTIVITY_CPU_THRESHOLD` => `CPU_ACTIVITY_THRESHOLD`, again it's not just about RStudio anymore. Also this is consistent to [what we have in the Helm chart](https://github.com/ministryofjustice/analytics-platform-helm-charts/blob/master/charts/idler/values.yaml#L21). (helm chart need updating to set the right env. If not it would default to 90% so not that breaking really.)
- [Stop](https://github.com/ministryofjustice/analytics-platform-idler/commit/a37ca72e7b40153b6d9e0a48646be8eda704aec1#diff-225132354aa2a9f4d7ecebb48833c06bR168) having the number of replicas in the `idled-at` annotation, this annotation will now only have the time when the app was idled. Simple and clear. (unidler or whoever uses this needs to change. **NOTE**: unidler is already robust enough to handle when this is not there or it's not valid)

Non-breaking:
- [added](https://github.com/ministryofjustice/analytics-platform-idler/commit/a37ca72e7b40153b6d9e0a48646be8eda704aec1#diff-225132354aa2a9f4d7ecebb48833c06b) `replicas-when-unidled` annotation with the sole responsibility of keeping the number of replicas desired at the time of idling.
- [got rid](https://github.com/ministryofjustice/analytics-platform-idler/commit/0b879b1196aa2d8ad31b415ebbf537c656c0858e) of unnecessary method `label_selector()` and moved reading of `LABEL_SELECTOR` env var at top of file (read inputs at the beginning of file/methods improves readability and it makes easier for someone to know what env variables are involved in the idling process)
- [updated](https://github.com/ministryofjustice/analytics-platform-idler/pull/68/commits/17b4c9267593a22a27373d43bce825b02a428e87) header comment.
- [bumped Python version](https://github.com/ministryofjustice/analytics-platform-idler/pull/68/commits/2d3e8de1e45bd8cef7e735941d5e53a3aa9bc73d) from `3.6.4` => `3.6.8`
- various `Dockerfile` improvements/cleanups:
  - [don't run as root](https://github.com/ministryofjustice/analytics-platform-idler/pull/68/commits/3a1705154c3cb2f3e6f5b3a8815b293175919326)
  - use `LABEL` instead of deprecated `MAINTAINER` command
  - use `COPY` instead of `ADD` (which can be sneaky)
  - grouped commands to install dependencies into a single `RUN` command/layer
- [updated](https://github.com/ministryofjustice/analytics-platform-idler/pull/68/commits/16791458010020efe9c48fb2378d0d8decadde18) test dependencies

### CODE REVIEW

**Comment by xoen** [comment]
Have a look but I need to update tests (apparently they don't automatically run in GH as well so probably another thing I'll have a look at).

**Comment by xoen** [comment]
@andyhd / @r4vi have a look please 😇

**Comment by xoen** [comment]
FYI: https://github.com/ministryofjustice/analytics-platform-idler/pull/68/commits/f75d69e10d3fc308901a6d8c639323837c88c157

File: Dockerfile [hunk 1]
```diff
@@ -7,6 +7,7 @@ RUN apk update && apk add --virtual build-dependencies build-base gcc libffi-dev
 
 ADD requirements.txt requirements.txt
 RUN pip install -U pip && pip install -r requirements.txt
```
Comment (r4vi) [inline Dockerfile]: this could be a single RUN with && now

File: Dockerfile [hunk 2]
```diff
@@ -7,6 +7,7 @@ RUN apk update && apk add --virtual build-dependencies build-base gcc libffi-dev
...
+RUN apk del build-dependencies
 
 ADD idler.py idler.py
 ADD metrics_api.py metrics_api.py
```
Comment (andyhd) [inline Dockerfile]: Should these be changed to `COPY` and combined into a single line?

File: Dockerfile [hunk 3]
```diff
@@ -7,6 +7,7 @@ RUN apk update && apk add --virtual build-dependencies build-base gcc libffi-dev
```
Comment (andyhd) [inline Dockerfile]: Also, `MAINTAINER` is deprecated in favour of `LABEL maintainer="email@address"`

File: idler.py [hunk 4]
```diff
@@ -1,8 +1,12 @@
 """
-Checks all RStudio deployments and idles those matching a kubernetes label
-selector.
-The label selector can be overridden by setting the LABEL_SELECTOR environment
-variable.
+Idles an application to save resources.
+
+Idling is performed by scaling down its deployment to zero replicas (no pods
```
Comment (andyhd) [inline idler.py]: Change `its` to `a`

File: Dockerfile (hunk 1 above)
Comment (xoen) [reply Dockerfile]: To be honest I've never been of dumping more and more stuff in a single line as it makes things dirty/hard to follow and I don't really see the benefit of it as you have few more small layers.

What would be the benefit of it?

Comment (r4vi) [reply Dockerfile]: in this case very little benefit, just 1 fewer layer but in general it's good to keep as many things in a single layer as possible. 

If you'd done this instead: 
```dockerfile
ADD requirements.txt requirements.txt

RUN apk update \
  && apk add --virtual build-dependencies build-base gcc libffi-dev openssl-dev \
  && pip install -U pip && pip install -r requirements.txt \
  && apk del build-dependencies
```
then the apk deps would never even be persisted to the layers in the image

File: Dockerfile (hunk 3 above)
Comment (xoen) [reply Dockerfile]: 👍 changing to:

```
LABEL maintainers="andy.driver@digital.justice.gov.uk,aldo.giambelluca@digital.justice.gov.uk"
```

File: idler.py (hunk 4 above)
Comment (xoen) [reply idler.py]: I'll change.

File: Dockerfile (hunk 2 above)
Comment (xoen) [reply Dockerfile]: I'll give it a shot 👍

File: Dockerfile (hunk 1 above)
Comment (xoen) [reply Dockerfile]: I'm having another look at the `Dockerfile` and conceptually these commands are run to achieve one single thing which is "install the dependencies" - therefore makes sense to bundle them all together.

Comment (xoen) [reply Dockerfile]: Also, now that I look at it. Even when the dependencies changes `apk update` etc... is not rerun as it would use the cached layer so you'll never get updated dependencies.

Whether that's a feature or a problem I do not know.

Comment (xoen) [reply Dockerfile]: I guess it's nice to have up-to-date libraries but if they break for whatever reason may be annoying. I suppose if I'm changing the `requirements.txt` to use a newer version of a module that may rely on a newer/shinier OS library so kinda makes sense to update them when updating the dependencies.

Comment (xoen) [reply Dockerfile]: FYI: https://github.com/ministryofjustice/analytics-platform-idler/pull/68/commits/7668a49fe53ee71bd76dd5460c16e631ab6fb68b
